Register keyboard shortcuts

miltonbsn opened the pull request:
# Summary

Fix some not intuitive behaviors related to canvas (Selection and drag) and add a set of new keyboard shortcuts.

# To do

`cmd` (Mac) == `ctrl` (Windows)
`opt` (Mac == `alt` (Windows)

- [x] Default mouse action to click+drag: Select elements
- [x] Hold `space bar` enable drag canvas mode
    - [x]  Hold `space bar` change cursor to `cursor: drag`
    - [x] Hold `space bar` + click change cursot to `cursor: dragging`
- [x] `ESC` key unselect element(s) (If any is selected)
- [x] Keyboard shortcut - `cmd` + `s` = Save model
- [x] Keyboard shortcut - `cmd` + `z` = undo last action
- [x] Keyboard shortcut - `cmd` + `shift` + `z` = redo previous undone action (If any)
- [x] Keyboard shortcut - `shift`+ `z` + `+` = Increase zoom by 10%
- [x] Keyboard shortcut - `shift` + `z` + `-` = Decrease zoom by 10%

# Related issues

- Close #77
- Close #208
- Close #209 
- Close #223 
- Close #233

miltonbsn committed:
Register shortcuts

idmarjr commented on app/angular/conceptual/conceptual.js:
```diff
@@ -317,11 +319,35 @@ const controller = function (ModelAPI, $stateParams, $rootScope, $timeout, $uibM
 			ctrl.showFeedback(false, "");
-			configs.selectionView.startSelecting(evt);
 			ctrl.onSelectElement(null);
+			configs.selectionView.cancelSelection();
```
Should we also `cancelSelection()` when user hit ESC key?

miltonbsn replied:
Good point ✅

miltonbsn committed:
Create keyboardController

idmarjr committed:
Add keyboard shortcut hint to .tool-icons items

Since we're not checking user OS in any ways, for now I'm using CTRL
as key identifier and assume Mac users will know they need to use CMD instead.

idmarjr committed:
Add keyboard shortcut hint to save button

miltonbsn committed:
Standardizes editors behavior

miltonbsn committed:
Merge branch 'feature/keyboard-shortcuts' of github.com:brmodeloweb/brmodelo-app into feature/keyboard-shortcuts

miltonbsn committed:
Remove missing debug log

idmarjr commented on app/joint/br-scroller.js:
```diff
@@ -118,19 +118,22 @@ joint.ui.PaperScroller = Backbone.View.extend({
-				var b = a.clientX - this._clientX,
-						c = a.clientY - this._clientY;
-				this.el.scrollTop -= c, this.el.scrollLeft -= b, this._clientX = a.clientX, this._clientY = a.clientY
+			$('.paper-scroller').css('cursor', 'grab');
```
Does it make sense to use vanilla ES2021 instead of jquery selector here? (Same for line 128)

Or we're stuck with jquery anyways?

CC @oicramps for an extra opinion 😁

miltonbsn replied:
jquery will live forever

ArthurMota9 reviewed (commented):
Great job Milton, users are sure to love these contributions. ❤️ 
I tested it here and it was very good. 🚀

In the KeyboardController class I have a suggestion here in this commit: https://github.com/brmodeloweb/brmodelo-app/commit/898e50372c42a4f3aa061d9e15124094920e1824
If you find it interesting I can pass this commit to your branch.

miltonbsn committed:
Add types and create hotkeys config (#253)

Authored-by: Arthur Mota <arthurmotaa9@gmail.com>

idmarjr reviewed (approved):
[Vai filhão](https://youtu.be/fJb_Jz_ce50)!

~~As soon as you answer my question here~~ 😛
